src/tiledbsoma_ml/dataset.py

```diff
@@ -5,218 +5,619 @@
-    and multi-process training such as :class:`torch.nn.parallel.DistributedDataParallel`, and will automatically partition
-    data appropriately. In the case of distributed training, sample partitions across all processes must be equal. Any
-    data tail will be dropped.
+    Shuffling support (in steps 2. and 3.) is enabled with the ``shuffle`` parameter, and should be used in lieu of
+    |DataLoader|'s default shuffling functionality. Similarly, |batch_size| should be used instead of |DataLoader|'s
+    default batching. |experiment_dataloader| is the recommended way to wrap an |ExperimentDataset| in a
+    |DataLoader|, as it enforces these constraints while passing through other |DataLoader| args.
+
+    Describing the whole process another way, we read randomly selected groups of ``obs`` coordinates from across all
+    |ExperimentAxisQuery| results, concatenate those into an I/O buffer, shuffle the buffer element-wise, fetch the full
+    row data (``X`` and ``obs``) for each coordinate, and send that on to PyTorch / the GPU, in batches. The randomness
+    of the shuffle is determined by:
+
+      - |shuffle_chunk_size|: controls the granularity of the global shuffle. ``shuffle_chunk_size=1`` corresponds to
+        a full global shuffle, but decreases I/O performance. Larger values cause chunks of rows to be shuffled,
+        increasing I/O performance (by taking advantage of data locality in the underlying |Experiment|) but decreasing
+        overall randomness of the yielded data.
+      - |io_batch_size|: number of rows to fetch at once (comprised of concatenated shuffle-chunks, and shuffled
+        row-wise). Larger values increase shuffle-randomness (by shuffling more "shuffle chunks" together), I/O
+        performance, and memory usage.
 
     Lifecycle:
         experimental
     """
 
+    layer_name: str
+    """``X`` layer to read (within an |ExperimentAxisQuery|'s |Measurement|)."""
+    obs_column_names: Sequence[str]
+    """Names of ``obs`` columns to return."""
+    batch_size: int
+    """Number of rows of ``X`` and ``obs`` data to yield in each |Batch|."""
+    io_batch_size: int
+    """Number of ``obs``/``X`` rows to fetch together, when reading from the provided |ExperimentAxisQuery|."""
+    shuffle: bool
+    """Whether to shuffle the ``obs`` and ``X`` data being returned."""
+    shuffle_chunk_size: int
+    """Number of contiguous rows shuffled as an atomic unit (before later concatenation/shuffling within IO-batches)."""
+    seed: int
+    """Random seed used for shuffling."""
+    return_sparse_X: bool
+    r"""When ``True``, return ``X`` data as a |csr_matrix| (by default, return |ndarray|\ s)."""
+    use_eager_fetch: bool
+    """Pre-fetch one "IO batch" and one "mini batch"."""
+
     def __init__(
         self,
         query: ExperimentAxisQuery,
-        X_name: str = "raw",
+        layer_name: str,
```

Comment:
breaking API change - needs to be noted in the CHANGELOG

(there are several such changes - a simple one-liner summarizing them is sufficient)